# Python 3.7

## Introduction:

* Python is a general-purpose interpreted, interactive, object-oriented, and high-level programming language
* Python is a high-level, interpreted, interactive and object-oriented scripting language. Python is designed to be highly readable
* It supports functional and structured programming methods as well as OOP.
* It can be used as a scripting language or can be compiled to byte-code for building large applications.
* It provides very high-level dynamic data types and supports dynamic type checking.
* It supports automatic garbage collection.

## New Features in Python3:

### The __future__ Module

Python 3 has some incompatiable keywords. For this The `__future__` module is introduced in Python 2.7, if incase you are planning to migrate to python 3. If incase you need python 3.x integer division in python2.7, use this import statement.

`from __future__ import division`

### The Print Statement

The Print Statement is now made with parenthesis '()' compulsory.
* `print "x"` --> is possible in python 2.x and not in python 3.x
* `print("x")` --> is possible in python 2.x and python 3.x
The print function in python 2.x introduces newline at the end of printed data by default. This can be supressed by adding `,` at the end of the statement.
* `print "x",` --> Supress newline character and prints `x`
In Python 3.x you need to specify a argument to the print function as `end=''` for no newline, `end=' '` for space instead of newline, `end="some_random_string"` for `some_random_string` instead of newline.
* `print("x", end=" ")` --> prints `x `
* `print("x", end="*")` --> prints `x*`
* `print("x", end="\n")` --> prints `x` with cursor at newline.

`_Python_3_Code_`

In [3]:
print("Kondiparthi Shanmukha Sarath", end="*")
print("is author of this notebook")

Kondiparthi Shanmukha Sarath*is author of this notebook


In [4]:
print("Kondiparthi Shanmukha Sarath", end="\n")
print("is author of this notebook")

Kondiparthi Shanmukha Sarath
is author of this notebook


In [7]:
print("Kondiparthi Shanmukha Sarath")
print("is author of this notebook")

Kondiparthi Shanmukha Sarath
is author of this notebook


In [8]:
print("Kondiparthi Shanmukha Sarath", end="")
print("is author of this notebook")

Kondiparthi Shanmukha Sarathis author of this notebook


### Reading Input

In Python 2.x there are two functions. `input() and raw_input()`.
* if `input()` is provided with numbers, it considers it as integer.
* if `input()` is provided with alphabets enclosed with `'` or `"`, it considers it as string.
* if `input()` is provided with numbers enclosed with `'` or `"`, it considers it as string.
* if `input()` is provided with alphabets (not enclosed with `'` or `"`), it considers as a declared variable name.
* if `raw_input()` is provided with either alphabets or numbers, which are either enclosed or not enclosed, by either `'` or `"`, it always considers them as string inputs.

In Python 3.x there is only one function. `input()`
* if `input()` is provided with either alphabets or numbers, which are either enclosed or not enclosed, by either `'` or `"`, it always considers them as string inputs.

Efficiently `input()` in python 2.x is deprecated in python 3.x and `raw_input()` in python 2.x is renamed as `input()` in python 3.x

`_Python_3_Code_`

In [12]:
input("Give any input, I read as String: ")

Give any input, I read as String: Kondiparthi Shanmukha Sarath is author of this book


'Kondiparthi Shanmukha Sarath is author of this book'

In [13]:
input("Give any input, I read as String: ")

Give any input, I read as String: 0123456789


'0123456789'

### Integer Division

 In Python 2.x | In Python 3.x
 ------------ | -----------
 `3/2 = 1` | `3/2 = 1.5`
 `3.0/2 = 1.5` | `3.0/2 = 1.5`
 `3/2.0 = 1.5` | `3/2.0 = 1.5`
 `77/34 = 2` | `77/34 = 2.264705882352941`
 `77.0/34 = 2.264705882352941` | `77.0/34 = 2.264705882352941`

`_Pyhton_3_Code_`

In [14]:
77/34

2.264705882352941

### Unicode Representation

* Python 2.x requires you to mark string as u incase you want that string to be stored as unicode.
* Python 3.x stores the string as unicode by default. (UTF-8)

### xrange() Function Removed

In Python 2.x | In Python 3.x
------------ | ----------
`range()` returns list | deprecated 
`xrange()` returns object that... | `xrange()` is renamed as `range()`...
...will only generate the list... | ...And even slicing is supported...
...of items in range on demand... | ...in python 3.2 and above.
... hence saving memory |  

### Raising Exceptions
It is recommended that we use parenthesis around the arguments of exception, while raising exceptions in python 3.x
* `raise IOError, "file error"` --> This is accepted in Python 2
* `raise IOError("file error")` --> This is also accepted in Python 2
* `raise IOError, "file error"` --> syntax error is raised in Python 3
* `raise IOError("file error")` --> This is the recommended syntax in Python 3

### Arguments in Exceptions
In Python 3, arguments to exception should be declared with 'as' keyword.
* `except Myerror, err:` --> In Python2
* `except Myerror as err:` --> In Python 3

### next() Function and .next() Method
`gen = (letter for letter in 'Hello World')`
In Python 2.x Interpretor
* `next(gen)` -> 'H'
* `gen.next()` -> 'e'
* `next(gen)` -> 'l'
* `gen.next()` -> 'l'
In Python 3.x Interpretor
* `next(gen)` -> 'H'
* `gen.next()` -> 'raises AttributeError'

In Python 3.x next() is no longer a generator method.

### 2to3 Utility
Along with Python 3 interpreter, 2to3.py script is usually installed in tools/scripts folder. It reads Python 2.x source code and applies a series of fixers to transform it into a valid Python 3.x code.

## Basic Syntax
### Style
* Interactive Mode
  * Open Python Interpretor by typing `Python` in terminal.
  * Run python commands diretly from the Interpretor. Example: `print("Hello World")`
* Script Mode
  * Enter all the python code (Example: `print("Hello World")`) in a file with `.py` extension. For example `any_name.py`
  * Type `python any_name.py`

### Python Identifiers
A Python identifier is a name used to identify a variable, function, class, module or other object. An identifier starts with a letter A to Z or a to z or an underscore (_) followed by zero or more letters, underscores and digits (0 to 9).

Python does not allow punctuation characters such as @, $, and % within identifiers. Python is a case sensitive programming language. Thus, Manpower and manpower are two different identifiers in Python.

* Class names start with an uppercase letter. All other identifiers start with a lowercase letter.
* Starting an identifier with a single leading underscore indicates that the identifier is private.
* Starting an identifier with two leading underscores indicates a strong private identifier.
* If the identifier also ends with two trailing underscores, the identifier is a language-defined special name.

### Python Reserved Keywords
____ | ___ | ___ | ___ | ___ | ___
----- | ---- | ---- | ---- | ---- | ----
and | exec | not | assert | for | pass 
as | finally | or | break | from | print
class | global | raise | continue | if | return
def | import | try | del | in | while
elif | is | with | else | lambda | yield
except |

### Lines and Indentation
* Python does not use braces({}) to indicate blocks of code for class and function definitions or flow control. Blocks of code are denoted by line indentation, which is rigidly enforced.

* The number of spaces in the indentation is variable, but all statements within the block must be indented the same amount.

### Multi-Line Statements
Statements in Python typically end with a new line. Python, however, allows the use of the line continuation character (\) to denote that the line should continue.

* `var1 = var2 + \`
* `       var3 + \`
* `       var4`

The statements contained within the [], {}, or () brackets do not need to use the line continuation character.

* `list23 = [1, 2, 3, 4,`
*             `5, 7]`

In [18]:
var1 = 2 + \
3 + \
4
print(var1)

9


In [19]:
list23 = [1, 2, 3, 4,
5, 7]
print(list23)

[1, 2, 3, 4, 5, 7]


### String quotations in Python
* Start with `'`, `"`, `'''` or `"""` and end with `'`, `"`, `'''` or `"""` respectively.
* `'` and `"` used when the strings are single lined.
* `'''` and `"""` are used strings are multilined.

In [22]:
string1 = 'This is a single lined string with single quotes'
string2 = "This is a single lined string with double quotes"
string3 = '''This is a multi-lined string wit
h three single quotes'''
string4 = """This is a multi-lined string with 
three double quotes"""
print (string1)
print (string2)
print (string3)
print (string4)

This is a single lined string with single quotes
This is a single lined string with double quotes
This is a multi-lined string wit
h three single quotes
This is a multi-lined string with 
three double quotes


In [24]:
string_that_gives_error_1 = 'This string is not single lined and enclosed with 
single quotes, hence this gives error'

SyntaxError: EOL while scanning string literal (<ipython-input-24-8930ec64f7e3>, line 1)

In [25]:
string_that_gives_error_2 = "This string is not single lined and enclosed with 
double quotes, hence this gives error"

SyntaxError: EOL while scanning string literal (<ipython-input-25-1e1f87463ad9>, line 1)

### Comments in Python
`#` is used for commenting.

**Examples**
* `# This is comment.`
* var1 = 48 `# This is also comment`
* `"""This is a multi-line string,
that can be used as multiline comments.
A String created and not assigned to any variable"""`

### Using Blank Lines
* A line containing only whitespace, possibly with a comment, is known as a blank line and Python totally ignores it.
* In an interactive interpreter session, you must enter an empty physical line to terminate a multiline statement.

### Multiple Statements on a Single Line
The semicolon ( ; ) allows multiple statements on a single line given that no statement starts a new code block.
* `import sys; x = 'foo'; sys.stdout.write(x + '\n')`

### Command Line Arguments

Save the Above Code in fine named test.py

Run the bash command as shown below.

> GetOpt module is also helpful in dealing with Command Line Arguments.
> **Refer getopt module docs.**

## Variables Types
### Assigning values
* Variables are nothing but reserved memory locations to store values. It means that when you create a variable, you reserve some space in the memory.
* Based on the data type of a variable, the interpreter allocates memory and decides what can be stored in the reserved memory. Therefore, by assigning different data types to the variables, you can store integers, decimals or characters in these variables.
* Python variables do not need explicit declaration to reserve memory space. The declaration happens automatically when you assign a value to a variable.

In [43]:
counter = 100
floating = 100.004
string_of_char = "shammy the raja"
print(counter)
print(floating)
print(string_of_char)

100
100.004
shammy the raja


### Multiple Assignment
* `a = b = c = 67`
* `a, b, c = 1, 2.3, 'shanmukha'`

In [49]:
a = b = c = 55
A, B, C = 1, 2.3, 'shanmukha'

### Standard Data Types
* Numbers
* String
* List
* Tuple
* Dictionary

### Python Numbers
Number data types store numeric values. Number objects are created when you assign a value to them
* `var1 = 29`
* `var2 = 44`
* `var3 = 33`
* `del var1`
* `del var2, var3`

There are three types of Numbers
* int (Singned integers)
* float (Floating Point Numbers)
* complex (Complex numbers)

> All integers in Python3 are represented as long integers. Hence, there is no separate number type as long.

### Python Strings
Please find the examples.

In [93]:
String_1 = "Kondiparthi Shanmukha Sarath is the author of this Notebook"
print(len(String_1))
print(String_1)                            # Prints all the letters in String
print(String_1[:])                         # Prints all the letters in String
print(String_1[-40:-7])                    # Prints letters from index = -40
print(String_1[3:37])                      # Prints letters from index = 3 to index = 36
print(String_1[37:3:-1])                   # -1 at the end gives the sring in reverse order
print(String_1[::-1])                      # Change parameter to -1, -2 -3 ... to draw insights
print(String_1 * 3)                        # Repetition
print(String_1 + "\n----THE END-------")   # String concatenantion

59
Kondiparthi Shanmukha Sarath is the author of this Notebook
Kondiparthi Shanmukha Sarath is the author of this Notebook
ha Sarath is the author of this N
diparthi Shanmukha Sarath is the a
ua eht si htaraS ahkumnahS ihtrapi
koobetoN siht fo rohtua eht si htaraS ahkumnahS ihtrapidnoK
Kondiparthi Shanmukha Sarath is the author of this NotebookKondiparthi Shanmukha Sarath is the author of this NotebookKondiparthi Shanmukha Sarath is the author of this Notebook
Kondiparthi Shanmukha Sarath is the author of this Notebook
----THE END-------


In [97]:
String_1[-59]                              # Try with different number as parameter

'K'

### Python Lists
Iterable, ordered and mutable Collection data.

In [106]:
var_list = [1, 3, 5, 212, 55]
# Converting the string of characters into list of characters.
list_of_chars = [x for x in (String_1)]
print(list_of_chars)

['K', 'o', 'n', 'd', 'i', 'p', 'a', 'r', 't', 'h', 'i', ' ', 'S', 'h', 'a', 'n', 'm', 'u', 'k', 'h', 'a', ' ', 'S', 'a', 'r', 'a', 't', 'h', ' ', 'i', 's', ' ', 't', 'h', 'e', ' ', 'a', 'u', 't', 'h', 'o', 'r', ' ', 'o', 'f', ' ', 't', 'h', 'i', 's', ' ', 'N', 'o', 't', 'e', 'b', 'o', 'o', 'k']


In [109]:
print(len(list_of_chars))
print(list_of_chars)                            # Prints all the letters in List
print("------------------------")
print(list_of_chars[:])                         # Prints all the letters in List
print("------------------------")
print(list_of_chars[-40:-7])                    # Prints letters from index = -40
print("------------------------")
print(list_of_chars[3:37])                      # Prints letters from index = 3 to index = 36
print("------------------------")
print(list_of_chars[37:3:-1])                   # -1 at the end gives the List in reverse order
print("------------------------")
print(list_of_chars[::-1])                      # Change parameter to -1, -2 -3 ... to draw insights
print("------------------------")
print(list_of_chars * 3)                        # Repetition
print("------------------------")
print(list_of_chars + ["appending", "another", "list"])   # List concatenantion

59
['K', 'o', 'n', 'd', 'i', 'p', 'a', 'r', 't', 'h', 'i', ' ', 'S', 'h', 'a', 'n', 'm', 'u', 'k', 'h', 'a', ' ', 'S', 'a', 'r', 'a', 't', 'h', ' ', 'i', 's', ' ', 't', 'h', 'e', ' ', 'a', 'u', 't', 'h', 'o', 'r', ' ', 'o', 'f', ' ', 't', 'h', 'i', 's', ' ', 'N', 'o', 't', 'e', 'b', 'o', 'o', 'k']
------------------------
['K', 'o', 'n', 'd', 'i', 'p', 'a', 'r', 't', 'h', 'i', ' ', 'S', 'h', 'a', 'n', 'm', 'u', 'k', 'h', 'a', ' ', 'S', 'a', 'r', 'a', 't', 'h', ' ', 'i', 's', ' ', 't', 'h', 'e', ' ', 'a', 'u', 't', 'h', 'o', 'r', ' ', 'o', 'f', ' ', 't', 'h', 'i', 's', ' ', 'N', 'o', 't', 'e', 'b', 'o', 'o', 'k']
------------------------
['h', 'a', ' ', 'S', 'a', 'r', 'a', 't', 'h', ' ', 'i', 's', ' ', 't', 'h', 'e', ' ', 'a', 'u', 't', 'h', 'o', 'r', ' ', 'o', 'f', ' ', 't', 'h', 'i', 's', ' ', 'N']
------------------------
['d', 'i', 'p', 'a', 'r', 't', 'h', 'i', ' ', 'S', 'h', 'a', 'n', 'm', 'u', 'k', 'h', 'a', ' ', 'S', 'a', 'r', 'a', 't', 'h', ' ', 'i', 's', ' ', 't', 'h', 'e', ' '

In [111]:
list_of_chars[-59]                                      # Try with different number input

'K'

### Python Tuples
Iterable, ordered and Immutable collection of data.

In [113]:
var_tuple = (1, 3, 5, 212, 55)
# Converting the string of characters into list of characters.
tuple_of_chars = tuple([x for x in (String_1)])
print(tuple_of_chars)

('K', 'o', 'n', 'd', 'i', 'p', 'a', 'r', 't', 'h', 'i', ' ', 'S', 'h', 'a', 'n', 'm', 'u', 'k', 'h', 'a', ' ', 'S', 'a', 'r', 'a', 't', 'h', ' ', 'i', 's', ' ', 't', 'h', 'e', ' ', 'a', 'u', 't', 'h', 'o', 'r', ' ', 'o', 'f', ' ', 't', 'h', 'i', 's', ' ', 'N', 'o', 't', 'e', 'b', 'o', 'o', 'k')


In [114]:
print(len(tuple_of_chars))
print(tuple_of_chars)                            # Prints all the letters in Tuple
print("------------------------")
print(tuple_of_chars[:])                         # Prints all the letters in Tuple
print("------------------------")
print(tuple_of_chars[-40:-7])                    # Prints letters from index = -40
print("------------------------")
print(tuple_of_chars[3:37])                      # Prints letters from index = 3 to index = 36
print("------------------------")
print(tuple_of_chars[37:3:-1])                   # -1 at the end gives the Tuple in reverse order
print("------------------------")
print(tuple_of_chars[::-1])                      # Change parameter to -1, -2 -3 ... to draw insights
print("------------------------")
print(tuple_of_chars * 3)                        # Repetition
print("------------------------")
print(tuple_of_chars + ("appending", "another", "list"))   # Tuple concatenantion

59
('K', 'o', 'n', 'd', 'i', 'p', 'a', 'r', 't', 'h', 'i', ' ', 'S', 'h', 'a', 'n', 'm', 'u', 'k', 'h', 'a', ' ', 'S', 'a', 'r', 'a', 't', 'h', ' ', 'i', 's', ' ', 't', 'h', 'e', ' ', 'a', 'u', 't', 'h', 'o', 'r', ' ', 'o', 'f', ' ', 't', 'h', 'i', 's', ' ', 'N', 'o', 't', 'e', 'b', 'o', 'o', 'k')
------------------------
('K', 'o', 'n', 'd', 'i', 'p', 'a', 'r', 't', 'h', 'i', ' ', 'S', 'h', 'a', 'n', 'm', 'u', 'k', 'h', 'a', ' ', 'S', 'a', 'r', 'a', 't', 'h', ' ', 'i', 's', ' ', 't', 'h', 'e', ' ', 'a', 'u', 't', 'h', 'o', 'r', ' ', 'o', 'f', ' ', 't', 'h', 'i', 's', ' ', 'N', 'o', 't', 'e', 'b', 'o', 'o', 'k')
------------------------
('h', 'a', ' ', 'S', 'a', 'r', 'a', 't', 'h', ' ', 'i', 's', ' ', 't', 'h', 'e', ' ', 'a', 'u', 't', 'h', 'o', 'r', ' ', 'o', 'f', ' ', 't', 'h', 'i', 's', ' ', 'N')
------------------------
('d', 'i', 'p', 'a', 'r', 't', 'h', 'i', ' ', 'S', 'h', 'a', 'n', 'm', 'u', 'k', 'h', 'a', ' ', 'S', 'a', 'r', 'a', 't', 'h', ' ', 'i', 's', ' ', 't', 'h', 'e', ' '

### Python Dictionary
Mutable Hash table, consisting unordered data in key-value pairs.

In [115]:
var_dict = {"initial_Key": 678, "some_Key": 78333.7834}
var_dict['one'] = "This is one"
var_dict[2] = "This is two"
var_dict["porto rico"] = "This is a city name"
print(var_dict)

{'initial_Key': 678, 'some_Key': 78333.7834, 'one': 'This is one', 2: 'This is two', 'porto rico': 'This is a city name'}


In [119]:
print(var_dict['one'])
print(var_dict['porto rico'])
print(var_dict.keys())
print(var_dict.values())
print(type(var_dict.keys()))

This is one
This is a city name
dict_keys(['initial_Key', 'some_Key', 'one', 2, 'porto rico'])
dict_values([678, 78333.7834, 'This is one', 'This is two', 'This is a city name'])
<class 'dict_keys'>


* `Dictionaries have no concept of order among the elements. It is incorrect to say that the elements are "out of order"; they are simply unordered.`

### Data Type Conversion
* `int(x, base=10)`           ----> Converts x to an integer. The base specifies the base if x is a string.
* `float(x)`                  ----> Converts x to a floating-point number.
* `complex(real, imag=0)`     ----> Creates a complex number.
* `str(x)`                    ----> Converts object x to a string representation.
* `repr(x)`                   ----> Converts object x to an expression string.
* `eval(string)`              ----> Evaluates a string and returns an object.
* `tuple(s)`                  ----> Converts s to a tuple.
* `list(s)`                   ----> Converts s to a List.
* `dict(d)`                   ----> Creates a dictionary. d must be a sequence of (key,value) tuples.
* `set(s)`                    ----> Converts s to a set.
* `frozenset(s)`              ----> Converts s to a frozen set.
* `chr(x)`                    ----> Converts an integer to a character.
* `unichr(x)`                 ----> Converts an integer to a Unicode character.
* `ord(x)`                    ----> Converts a single character to its integer value.
* `hex(x)`                    ----> Converts an integer to a hexadecimal string.
* `oct(x)`                    ----> Converts an integer to an octal string.

Experiment them.

In [125]:
ord('C')

67

## Basic Operators
### Types of Operators
* Arithematic Operators
* Comparison (Relational) Operators
* Assignment Operators
* Logical Operators
* Bitwise Operators
* Membership Operators
* Identity Operators

### Python Arithematic Operators
`+Addition, -Subtraction, *Multiplication, /Division, %Modulus, **Exponent, //Floor Division`

Experiment them.

In [130]:
a=23.456
b=5
print(a//b)

4.0


### Python Comparison Operators
`==Equality, !=NegationOfEquality, >GreaterThan, <LessThan, >=GreaterThanorEqualsTo, <=LessThanOrEqualsTo`

Experiment them

In [135]:
a=23.456     # a='kondiparthi shanmukha sarath'
b=5          # b='Ponnapalli Thanuja'
print(a<b)   # Try with strings too....

False


### Python Assignment Operators
`=Assign, +=AddAndAssign, -=SubtractAndAssign, *=MultiplyAndAssign, /=DivideAndAssign, %=ModulusAndAssign, **=ExponentAndAssign, //=FloorDivisionAndAssign`

Experiment them.

In [137]:
a=23.456
b=5
a+=b
print(a)

28.456


### Python Bitwise Operators
`&AND, |OR, ^XOR, ~Compliment, <<LeftShift, >>RightShift`

Experiment them.

In [142]:
a=60
b=13
print(bin(a))
print(bin(b))
print(bin(a | b))
print(a | b)

0b111100
0b1101
0b111101
61


### Python Logical Operators
`and, or, not`

Experiment them.

In [143]:
a=True
b=False
print(a and b)

False


### Python Membership Operators
`in, not in`

Python’s membership operators test for membership in a sequence, such as strings, lists, or tuples.

Experiment them

In [148]:
a = 45
b = [45, 67, 88, 22]
print(a in b)

True


### Python Identity Operators
`is, is not`

Identity operators compare the memory locations of two objects.

Experiment them.

In [153]:
a = c ='some_string'
b = 'some_other_string'
d = 7
e = 73
f = [1, 2, 3, 4]
g = [1, 2, 3, 4]
print(a is b)
print(a in c)
print(d is e)
print(f is g)

False
True
False
False


### Python Operators Precedence
No one cares.

## Python 3 - Decision Making.
### If else statements.
Here Sample skeleton of the if-else structure. Experiment with different types of data in condition part of if, elif.

In [166]:
condition = 0
#condition = -33
#condition = 22
#condition = True
#condition = 30 in [20, 30, 50]
if condition:
    print('KondiparthiShanmukhaSarath')

## Python 3 - Loops.
### Types of Loops.
* while loop
* for loop
* nested loop

### While Loop

In [168]:
count = 0
while (count < 9):
   print ('The count is:', count)
   count = count + 1
else:
    print("This is printed only if while loop is exited normally, and not by using any break statement.")
print ("Good bye!")

The count is: 0
The count is: 1
The count is: 2
The count is: 3
The count is: 4
The count is: 5
The count is: 6
The count is: 7
The count is: 8
This is printed only if while loop is exited normally, and not by using any break statement.
Good bye!


#### Infinite while loop

In [170]:
flag = 0            # change flag = 1 to start infinite while loop.
while (flag): print ('Given flag is really true!')
print ("Good bye!")

Good bye!


### For Loop

Here is the skeleton for For loop.

#### The range() function
range() function can be used to generate sequence objects. ie. these objects can be iterated.

In [3]:
range(0, 6)

range(0, 6)

In [4]:
list(range(0, 6))

[0, 1, 2, 3, 4, 5]

In [10]:
for letter in 'Python':     # traversal of a string sequence
   print ('Current Letter :', letter)
print()

fruits = ['banana', 'apple',  'mango']
for fruit in fruits:        # traversal of List sequence
   print ('Current fruit :', fruit)

list_of_items = range(0, 8)
for item in list_of_items:
    print (item, end=" ")

print ("Good bye!")

Current Letter : P
Current Letter : y
Current Letter : t
Current Letter : h
Current Letter : o
Current Letter : n

Current fruit : banana
Current fruit : apple
Current fruit : mango
0 1 2 3 4 5 6 7 Good bye!


In [13]:
for item in list_of_items:
    print (item, end=" ")
    #break
else:
    print ("This statement is executed only if for loop terminates natural exhaustion, not by break statement")

0 1 2 3 4 5 6 7 This statement is executed only if for loop terminates natural exhaustion, not by break statement


### Nested Loops
Skeletons for nested loops

### Loop Control Statements
* `break` statement
* `continue` statement
* `pass` statement

`break` is used to stop the iteration of loops and continue the execution of statements right after the loop. `continue` is used to skip the current iteration of loop statements and proceed to next iteration.
`pass` is used when you syntactically require the statement, but you dont want to have any code to be executed.

### Iterator and Generator
terator is an object which allows a programmer to traverse through all the elements of a collection, regardless of its specific implementation. In Python, an iterator object implements two methods, iter() and next().

String, List or Tuple objects can be used to create an Iterator.

A generator is a function that produces or yields a sequence of values using yield method.

When a generator function is called, it returns a generator object without even beginning execution of the function. When the next() method is called for the first time, the function starts executing until it reaches the yield statement, which returns the yielded value. The yield keeps track i.e. remembers the last execution and the second next() call continues from previous value.

In [29]:
list_of_numbers = list(range(4,13))
print("list_of_numbers: " + str(list_of_numbers))
iter_items = iter(list_of_numbers)
#Either for loop or while loop.
#for x in iter_items:
    #print(x, end=" ")
while True:
    try:
        print (next(iter_items), end=", ")
    except StopIteration:
        print ("Iteration is done...")
        break

list_of_numbers: [4, 5, 6, 7, 8, 9, 10, 11, 12]
4, 5, 6, 7, 8, 9, 10, 11, 12, Iteration is done...


In [30]:
def fibonacci(n): #generator function
    a, b, counter = 0, 1, 0
    while True:
        if (counter > n): 
            return
        yield a
        a, b = b, a + b
        counter += 1
        
f = fibonacci(5) #f is iterator object.
while True:
    try:
        print (next(f), end=", ")
    except StopIteration:
        print("Iteration is done...")
        break

0, 1, 1, 2, 3, 5, Iteration is done...


## Python 3 - Numbers
`int, float, complex (a+jb)`. 

### Mathematical Functions
`Import math` where ever required.

function | Description
---------| -----------
abs(x) | The absolute value of x: the (positive) distance between x and zero.
ceil(x) | The ceiling of x: the smallest integer not less than x.
cmp(x, y) | -1 if x < y, 0 if x == y, or 1 if x > y. Deprecated in Python 3. Use (x>y)-(x<y).
exp(x) | The exponential of x: e^x.
fabs(x) | The absolute value of x.
floor(x) | The floor of x: the largest integer not greater than x.
log(x) | The natural logarithm of x, for x > 0.
log10(x) | The base-10 logarithm of x for x > 0.
max(x1, x2, x3, x4, x5...) | The largest of its arguments: the value closest to positive infinity.
min(x1, x2, x3, x4, x5...) | The smallest of its arguments: the value closest to negative infinity.
(math.modf) modf(x) | The fractional and integer parts of x in a two-item tuple. Both parts have the same sign as x. The integer part is returned as a float.
pow(x, y) | x\*\*y
round(x, n=0) | x rounded to n digits from the decimal point. Python rounds away from zero as a tie-breaker: round(0.5) is 1.0 and round(-0.5) is -1.0.
sqrt(x) | The square root of x for x > 0.

In [48]:
math.log(5)

1.6094379124341003

### Random Number Functions
function | description
--------- | ----------
choice(seq) | A random item from a list, tuple, or string.
randrange ([start,] stop [,step]) | A randomly selected element from range(start, stop, step).
random() | A random float r, such that 0 is less than or equal to r and r is less than 1
seed([x]) | Sets the integer starting value used in generating random numbers. Call this function before calling any other random module function. Returns None.
shuffle(lst) | Randomizes the items of a list in place. Returns None.
uniform(x, y) | A random float r, such that x is less than or equal to r and r is less than y.

In [90]:
import random
#random.seed(3)
for i in [1, 2, 3, 4, 5, 6, 7, 8]:
    print(random.random())

0.5842517929701989
0.9042017708477751
0.6819821366349667
0.928945601200017
0.8564005663967557
0.9909896448688151
0.6712735421625182
0.16309962197106975


### Trigonometric Functions
https://www.tutorialspoint.com/python3/python_numbers.htm to check trignometric functions on `asin, acos, atan, sin, cos etc`.

* `degrees(x)` --> x from radians to degrees
* `radians(x)` --> x from degrees to radians

In [92]:
math.degrees(2)

114.59155902616465

### Mathematical Constants
* `math.pi`
* `math.e`

In [94]:
math.e

2.718281828459045

## Python 3 - Strings
### Escape Characters
* `\a` ---> Bell or alert
* `\b` ---> Backspace
* `\cx or \Cx` ---> Control-x
* `\e` ---> Escape
* `\f` ---> Formfeed
* `\n` ---> Newline
* `\r` ---> Return
* `\s` ---> Space
* `\t` ---> Tab
* `\v` ---> Vertical Tab

In [1]:
print("This is a example \n \t for escape characters")

This is a example 
 	 for escape characters


### Raw Strings

In [5]:
print(r'\n\t\s all these backslashes are not processed, ie not used as escape characters')

\n\t\s all these backslashes are not processed, ie not used as escape characters


### String Formatting Operator
* `%c` ---> character
* `%s` ---> string
* `%i` ---> signed decimal integer
* `%d` ---> signed decimal integer
* `%u` ---> unsigned decimal integer
* `%o` ---> octal integer
* `%x` ---> hexadecimal integer (lower case)
* `%X` ---> hexadecimal integer (upper case)
* `%e` ---> exponential notation (lower case e)
* `%E` ---> exponential notation (upper case E)
* `%f` ---> floating point
* `%g` ---> the shorter of %f and %e
* `%G` ---> the shorter of %f and %E

In [4]:
print ("My name is %s and weight is %d kg!" % ('Zara', 21)) 
print ("%X" % (678))

My name is Zara and weight is 21 kg!
2A6


**`ALL THE STRINGS IN PYTHON 3 IS UNICODE BY DEFAULT`**

### Butilt in String Methods
* `capitalize()` ---> Capitalizes the first letter of the string
* `center(width, fillchar)` ---> Returns a string padded with fillchar with the original string centered to a total of width columns.
* `count(str, beg=0, end=len(string))` ---> Counts how many times str occurs in string or in a substring of string if starting index beg and ending index end are given.
* `decode(encoding='UTF-8',errors='strict')` ---> Decodes the string using the codec registered for encoding. encoding defaults to the default string encoding.
* `encode(encoding='UTF-8',errors='strict')` ---> Returns encoded string version of string; on error, default is to raise a ValueError unless errors is given with 'ignore' or 'replace'.
* `endswith(suffix, beg=0, end=len(string))` ---> Determines if string or a substring of string (if starting index beg and ending index end are given) ends with suffix; returns true if so and false otherwise.
* `expandtabs(tabsize = 8)` ---> Expands tabs in string to multiple spaces; defaults to 8 spaces per tab if tabsize not provided.
* `find(str, beg = 0 end = len(string))` ---> Determine if str occurs in string or in a substring of string if starting index beg and ending index end are given returns index if found and -1 otherwise.
* `index(str, beg = 0, end = len(string))` ---> Same as find(), but raises an exception if str not found.
* `isalnum()` ---> Returns true if string has at least 1 character and all characters are alphanumeric and false otherwise.
* `isalpha()` ---> Returns true if string has at least 1 character and all characters are alphabetic and false otherwise.
* `isdigit()` ---> Returns true if string contains only digits and false otherwise.
* `islower()` ---> Returns true if string has at least 1 cased character and all cased characters are in lowercase and false otherwise.
* `isnumeric()` ---> Returns true if a unicode string contains only numeric characters and false otherwise.
* `isspace()` ---> Returns true if string contains only whitespace characters and false otherwise.
* `istitle()` ---> Returns true if string is properly "titlecased" and false otherwise.
* `isupper()` ---> Returns true if string has at least one cased character and all cased characters are in uppercase and false otherwise.
* `join(seq)` ---> Merges (concatenates) the string representations of elements in sequence seq into a string, with separator string.
* `len(string)` ---> Returns the length of the string
* `ljust(width[, fillchar])` ---> Returns a space-padded string with the original string left-justified to a total of width columns.
* `lower()` ---> Converts all uppercase letters in string to lowercase.
* `lstrip()` ---> Removes all leading whitespace in string.
* `maketrans()` ---> Returns a translation table to be used in translate function.
* `max(str)` ---> Returns the max alphabetical character from the string str.
* `min(str)` ---> Returns the min alphabetical character from the string str.
* `replace(old, new [, max])` ---> Replaces all occurrences of old in string with new or at most max occurrences if max given.
* `rfind(str, beg = 0,end = len(string))` ---> Same as find(), but search backwards in string.
* `rindex( str, beg = 0, end = len(string))` ---> Same as index(), but search backwards in string.
* `rjust(width,[, fillchar])` ---> Returns a space-padded string with the original string right-justified to a total of width columns.
* `rstrip()` ---> Removes all trailing whitespace of string.
* `split(str="", num=string.count(str))` ---> Splits string according to delimiter str (space if not provided) and returns list of substrings; split into at most num substrings if given.
* `splitlines( num=string.count('\n'))` ---> Splits string at all (or num) NEWLINEs and returns a list of each line with NEWLINEs removed.
* `startswith(str, beg=0,end=len(string))` ---> Determines if string or a substring of string (if starting index beg and ending index end are given) starts with substring str; returns true if so and false otherwise.
* `strip([chars])` ---> Performs both lstrip() and rstrip() on string
* `swapcase()` ---> Inverts case for all letters in string.
* `title()` ---> Returns "titlecased" version of string, that is, all words begin with uppercase and the rest are lowercase.
* `translate(table, deletechars="")` ---> Translates string according to translation table str(256 chars), removing those in the del string.
* `upper()` ---> Converts lowercase letters in string to uppercase.
* `zfill (width)` ---> Returns original string leftpadded with zeros to a total of width characters; intended for numbers, zfill() retains any sign given (less one zero).
* `isdecimal()` ---> Returns true if a unicode string contains only decimal characters and false otherwise.

In [30]:
"+shammy".zfill(16)

'+000000000shammy'

## Python 3 - Lists
The most basic data structure in Python is the sequence. Each element of a sequence is assigned a number - its position or index. The first index is zero, the second index is one, and so forth.

* `len(list_1)` ---> Gives total length of the list.
* `max(list_1)` ---> Returns item from the list with max value.
* `min(list_1)` ---> Returns item from the list with min value.
* `list(seq)` ---> Converts a tuple into list.
* `list_1.append(obj)` ---> Appends object obj to list.
* `list_1.count(obj)` ---> Returns count of how many times obj occurs in list.
* `list_1.extend(seq)` ---> Appends the contents of seq to list.
* `list_1.index(obj)` ---> Returns the lowest index in list that obj appears.
* `list_1.insert(index, obj)` ---> Inserts object obj into list at offset index
* `list_1.pop(obj = list_1[-1])` ---> Removes and returns last object or obj from list
* `list_1.remove(obj)` ---> Removes object obj from list
* `list_1.reverse()` ---> Reverses objects of list in place
* `list_1.sort([func])` ---> Sorts objects of list, use compare func if given

In [41]:
list_1 = [(2, 5), (3, "shanmmu", 98, True), ("thanuja", True, 9088)]
list_1.sort(key=lambda x:len(x), reverse=True)
print(list_1)

[(3, 'shanmmu', 98, True), ('thanuja', True, 9088), (2, 5)]


In [43]:
list_2 = [324, 8734, 897, 999]
max(list_2)

8734

## Python 3 - Tuples
A tuple is a sequence of immutable Python objects.

`tuple_1 = (34234, "shammy", True, "shiny", "Thanuja")`

`tuple_2 = ("This tuple has only one item, hence it requires comma after that one item",)`

`tuple_3 = ("Thanuja")`

Here tuple_3 is not tuple, but a string.

* `cmp(tuple1, tuple2)` ---> Compares elements of both tuples.
* `len(tuple)` ---> Gives the total length of the tuple.
* `max(tuple)` ---> Returns item from the tuple with max value.
* `min(tuple)` ---> Returns item from the tuple with min value.
* `tuple(seq)` ---> Converts a list into tuple.

In [47]:
tuple_1 = ("Thanuja",)
print(tuple_1)

('Thanuja',)


## Python 3 - Dictionary
Each key is separated from its value by a colon (:), the items are separated by commas, and the whole thing is enclosed in curly braces. An empty dictionary without any items is written with just two curly braces, like this: {}.

Keys are unique within a dictionary while values may not be. The values of a dictionary can be of any type, but the keys must be of an immutable data type such as strings, numbers, or tuples.

**`More than one entry per key is not allowed`**

**`Keys must be immutable`**

* `len(dict_1)` ---> Gives the total length of the dictionary. This would be equal to the number of items in the dictionary.
* `str(dict_1)` ---> Gives the total length of the dictionary. This would be equal to the number of items in the dictionary.
* `type(variable)` ---> Returns the type of the passed variable. If passed variable is dictionary, then it would return a dictionary type.
* `dict_1.clear()` ---> Removes all elements of dictionary dict
* `dict_1.copy()` ---> Returns a shallow copy of dictionary dict
* `dict_1.fromkeys(seq[, value]))` ---> Create a new dictionary with keys from seq and values set to value.
* `dict_1.get(key, default=None)` ---> For key key, returns value or default if key not in dictionary
* `dict_1.has_key(key)` ---> Removed, use the in operation instead.
* `dict_1.items()` ---> Returns a list of dict's (key, value) tuple pairs
* `dict_1.keys()` ---> Removed, use the in operation instead.
* `dict_1.setdefault(key, default = None)` ---> Similar to get(), but will set dict[key] = default if key is not already in dict
* `dict_1.update(dict2)` ---> Adds dictionary dict2's key-values pairs to dict
* `dict_1.values()` ---> Returns list of dictionary dict's values

In [49]:
dict_1 = {'Name': 'Zara', 'Age': 7, 'Class': 'First'}
dict_1['Shammy'] = "It is me"
print(dict_1)

{'Name': 'Zara', 'Age': 7, 'Class': 'First', 'Shammy': 'It is me'}


## Python 3 - Date & Time
Python's time and calendar modules help track dates and times.

20 tick - 1 second

Particular instants in time are expressed in seconds since 12:00am, January 1, 1970(epoch). The function time.time() returns the current system time in ticks since 12:00am, January 1, 1970(epoch). Date arithmetic is easy to do with ticks. However, dates before the epoch cannot be represented in this form. Dates in the far future also cannot be represented this way - the cutoff point is sometime in 2038 for UNIX and Windows.

In [7]:
import time;

localtime = time.localtime(time.time())
print(time.time())
print ("Local current time :" + str(localtime))
localtime = time.asctime(time.localtime(time.time()))
print(localtime)

1568185184.983774
Local current time :time.struct_time(tm_year=2019, tm_mon=9, tm_mday=11, tm_hour=12, tm_min=29, tm_sec=44, tm_wday=2, tm_yday=254, tm_isdst=0)
Wed Sep 11 12:29:44 2019


In [9]:
import calendar

cal = calendar.month(2019, 1)
print (cal)

    January 2019
Mo Tu We Th Fr Sa Su
    1  2  3  4  5  6
 7  8  9 10 11 12 13
14 15 16 17 18 19 20
21 22 23 24 25 26 27
28 29 30 31



# The Time Module Fucntions
* `time.asctime([tupletime])` ---> Accepts a time-tuple and returns a readable 24-character string such as 'Tue Dec 11 18:07:14 2008'.
* `time.clock()` ---> Returns the current CPU time as a floating-point number of seconds. To measure computational costs of different approaches, the value of time.clock is more useful than that of time.time().
* `time.ctime([secs])` ---> Like time.asctime()
* `time.gmtime([secs])` ---> Accepts an instant expressed in seconds since the epoch and returns a time-tuple t with the UTC time. Note : t.tm_isdst is always 0
* `time.localtime([secs])` ---> Accepts an instant expressed in seconds since the epoch and returns a time-tuple t with the local time (t.tm_isdst is 0 or 1, depending on whether DST applies to instant secs by local rules).
* `time.mktime(tupletime)` ---> Accepts an instant expressed as a time-tuple in local time and returns a floating-point value with the instant expressed in seconds since the epoch.
* `time.sleep(secs)` ---> Suspends the calling thread for secs seconds.
* `time.strftime(fmt[,tupletime])` ---> Accepts an instant expressed as a time-tuple in local time and returns a string representing the instant as specified by string fmt.
* `time.strptime(str,fmt='%a %b %d %H:%M:%S %Y')` ---> Parses str according to format string fmt and returns the instant in time-tuple format.
* `time.time( )` ---> Returns the current time instant, a floating-point number of seconds since the epoch.
* `time.tzset()` ---> Resets the time conversion rules used by the library routines. The environment variable TZ specifies how this is done.
* `time.timezone` ---> Attribute time.timezone is the offset in seconds of the local time zone (without DST) from UTC (>0 in the Americas; <=0 in most of Europe, Asia, Africa).
* `time.tzname` ---> Attribute time.tzname is a pair of locale-dependent strings, which are the names of the local time zone without and with DST, respectively.

In [22]:
time.ctime()

'Wed Sep 11 13:05:57 2019'

# The Calendar Module
By default, calendar takes Monday as the first day of the week and Sunday as the last one. To change this, call calendar.setfirstweekday() function.
* `calendar.calendar(year,w=2,l=1,c=6)` ---> 
* `calendar.firstweekday( )` ---> 
* `calendar.isleap(year)` ---> 
* `calendar.leapdays(y1,y2)` ---> 
* `calendar.month(year,month,w=2,l=1)` ---> 
* `calendar.monthcalendar(year,month)` ---> 
* `calendar.monthrange(year,month)` ---> 
* `calendar.prcal(year,w=2,l=1,c=6)` ---> 
* `calendar.prmonth(year,month,w=2,l=1)` ---> 
* `calendar.setfirstweekday(weekday)` ---> 
* `calendar.timegm(tupletime)` ---> 
* `calendar.weekday(year,month,day)` ---> 

In [42]:
import calendar
print (calendar.monthcalendar(2019 , 2))
print (calendar.leapdays(2000,2012))

[[0, 0, 0, 0, 1, 2, 3], [4, 5, 6, 7, 8, 9, 10], [11, 12, 13, 14, 15, 16, 17], [18, 19, 20, 21, 22, 23, 24], [25, 26, 27, 28, 0, 0, 0]]
3


## Python - Functions
Examples:

### Anonymous Functions.
`lambda [arg1 [,arg2,.....argn]]:expression`
These functions are called anonymous because they are not declared in the standard manner by using the def keyword. You can use the lambda keyword to create small anonymous functions.
* Lambda forms can take any number of arguments but return just one value in the form of an expression. They cannot contain commands or multiple expressions.
* An anonymous function cannot be a direct call to print because lambda requires an expression
* Lambda functions have their own local namespace and cannot access variables other than those in their parameter list and those in the global namespace.
* Although it appears that lambda's are a one-line version of a function, they are not equivalent to inline statements in C or C++, whose purpose is by passing function stack allocation during invocation for performance reasons.

In [44]:
sum_1 = lambda val_1, val_2 : val_1 + val_2
sum_1(3, 8)

11

## Python - Modules
A module allows you to logically organize your Python code. Grouping related code into a module makes the code easier to understand and use. A module is a Python object with arbitrarily named attributes that you can bind and reference.

Simply, a module is a file consisting of Python code. A module can define functions, classes and variables. A module can also include runnable code.

### The import Statement
You can use any Python source file as a module by executing an import statement in some other Python source file.

`import module1[, module2[,... moduleN]`

`from modname import name1[, name2[, ... nameN]]`

`from modname import *`

### Locating Modules
When you import a module, the Python interpreter searches for the module in the following sequences −

The current directory.

If the module isn't found, Python then searches each directory in the shell variable PYTHONPATH.

If all else fails, Python checks the default path. On UNIX, this default path is normally /usr/local/lib/python/.

The module search path is stored in the system module sys as the sys.path variable. The sys.path variable contains the current directory, PYTHONPATH, and the installation-dependent default.

### The PYTHONPATH Variable
The PYTHONPATH is an environment variable, consisting of a list of directories. The syntax of PYTHONPATH is the same as that of the shell variable PATH

`set PYTHONPATH = /usr/local/lib/python`

### Namespaces and Scoping

Variables are names (identifiers) that map to objects. A namespace is a dictionary of variable names (keys) and their corresponding objects (values).

A Python statement can access variables in a local namespace and in the global namespace. If a local and a global variable have the same name, the local variable shadows the global variable.

Each function has its own local namespace. Class methods follow the same scoping rule as ordinary functions.

Python makes educated guesses on whether variables are local or global. It assumes that any variable assigned a value in a function is local.

Therefore, in order to assign a value to a global variable within a function, you must first use the global statement.

The statement global VarName tells Python that VarName is a global variable. Python stops searching the local namespace for the variable.

For example, we define a variable Money in the global namespace. Within the function Money, we assign Money a value, therefore Python assumes Money as a local variable. However, we accessed the value of the local variable Money before setting it, so an UnboundLocalError is the result. Uncommenting the global statement fixes the problem.

In [51]:
money_1 = 200
def some_function():
    global money_1
    money_1 = money_1 + 1
    return money_1
    
some_var = some_function()
print(some_var)

201


### The dir( ) Function
The dir() built-in function returns a sorted list of strings containing the names defined by a module.

The list contains the names of all the modules, variables and functions that are defined in a module

In [52]:
import math

content = dir(math)
print (content)

['__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'copysign', 'cos', 'cosh', 'degrees', 'e', 'erf', 'erfc', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'pi', 'pow', 'radians', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc']


### The globals() and locals() Functions
* The globals() and locals() functions can be used to return the names in the global and local namespaces depending on the location from where they are called.
* If locals() is called from within a function, it will return all the names that can be accessed locally from that function.
* If globals() is called from within a function, it will return all the names that can be accessed globally from that function.
* The return type of both these functions is dictionary. Therefore, names can be extracted using the keys() function.

### The reload() Function
* When the module is imported into a script, the code in the top-level portion of a module is executed only once.
* Therefore, if you want to reexecute the top-level code in a module, you can use the reload() function. The reload() function imports a previously imported module again.
* `reload(module_name)`
* `reload(math)`

### Packages in Python
A package is a hierarchical file directory structure that defines a single Python application environment that consists of modules and subpackages and sub-subpackages, and so on.

## Python 3 - Files I/O
### print statement
You can pass zero or more arguments, the output will be a string conatenated.

In [4]:
print("Shammy", "Shiny", "Thanuja", "Sarma", "Sarada", sep="_____", end="*")

Shammy_____Shiny_____Thanuja_____Sarma_____Sarada*

### input statement
Reads any input inform of string. whether or not enclosed by "", ''.

In [7]:
some_var = input("Type something: ")
print(type(some_var))
print(some_var)

Type something: 97980"98098"
<class 'str'>
97980"98098"


### Opening and Closing Files
#### The open Function
`file object = open(file_name [, access_mode][, buffering])`

**file_name** − The file_name argument is a string value that contains the name of the file that you want to access.

**access_mode** − The access_mode determines the mode in which the file has to be opened, i.e., read, write, append, etc. A complete list of possible values is given below in the table. This is an optional parameter and the default file access mode is read (r).

**buffering** − If the buffering value is set to 0, no buffering takes place. If the buffering value is 1, line buffering is performed while accessing a file. If you specify the buffering value as an integer greater than 1, then buffering action is performed with the indicated buffer size. If negative, the buffer size is the system default(default behavior).

mode | pointer_position | overwrites_file | replace_data | appends | creates_new_file_if_not_available | reads_data | writes_data | binary | description
-----|------------------|-----------------|--------------|---------|-----------------------------------|-------|------------|--------|------------
r | beginning | No | No | No | No | Yes | No | No |
rb | beginning | No | No | No | No | Yes | No | Yes |
r+ | beginning | No | Yes | No | No | Yes | Yes | No |
rb+ | beginning | No | Yes | No | No | Yes | Yes | Yes |
w | beginning | Yes | No | No | Yes | No | Yes | No |
wb | beginning | Yes | No | No | Yes | No | Yes | Yes |
w+ | beginning | Yes | No | No | Yes | Yes | Yes | No |
wb+ | beginning | Yes | No | No | Yes | Yes | Yes | Yes |
a | end | No | No | Yes | Yes | No | Yes | No |
ab | end | No | No | Yes | Yes | No | Yes | Yes |
a+ | end | No | No | Yes | Yes | Yes | Yes | No |
ab+ | end | No | No | Yes | Yes | Yes | Yes | Yes |

In [19]:
file_object = open("/Users/apple/Desktop/go_green.txt", "a")

In [20]:
file_object.write("shammy")

6

In [21]:
file_object.close()

#### File Attributes
* `file.closed` ---> Returns true if file is closed, false otherwise.
* `file.mode` ---> Returns access mode with which file was opened.
* `file.name` ---> Returns name of the file.

### The close() Method
The close() method of a file object flushes any unwritten information and closes the file object, after which no more writing can be done.

Python automatically closes a file when the reference object of a file is reassigned to another file. It is a good practice to use the close() method to close a file.

`fileObject.close()`
### Reading and Writing Files
#### The write() Method
The write() method writes any string to an open file. It is important to note that Python strings can have binary data and not just text.

The write() method does not add a newline character ('\n') to the end of the string
`fileObject.write(string)`

In [25]:
file_object = open("/Users/apple/Desktop/go_green.txt", "r+")
file_object.write("Shiny")
file_object.close()

#### The read() Method
The read() method reads a string from an open file. It is important to note that Python strings can have binary data. apart from text data.

`fileObject.read([count])`

Here, passed parameter is the number of bytes to be read from the opened file. This method starts reading from the beginning of the file and if count is missing, then it tries to read as much as possible, maybe until the end of file.

In [26]:
file_object = open("/Users/apple/Desktop/go_green.txt", "r+")
file_data = file_object.read(10)
print(file_data)
file_object.close()

Shinyysham


### File Positions
* `tell()` ---> tells you the current position within the file
* `seek(offset[, from])` ---> changes the current file position, The offset argument indicates the number of bytes to be moved. The from argument specifies the reference position from where the bytes are to be moved. If from is set to 0, the beginning of the file is used as the reference position. If it is set to 1, the current position is used as the reference position. If it is set to 2 then the end of the file would be taken as the reference position.

### Renaming and Deleting Files
* `import os as os`
* `os.rename(current_file_name, new_file_name)`
* `os.remove(file_name)`

### Directories in Python
* `os.mkdir("newdir")`
* `os.chdir("newdir")`
* `os.getcwd()`
* `os.rmdir('dirname')`

## Python 3 - File Methods
* `file.open()` ---> Opens file to perform other operations listed below.
* `file.close()` ---> Close the file. A closed file cannot be read or written any more.
* `file.flush()` ---> Flush the internal buffer, like stdio's fflush. This may be a no-op on some file-like objects.
* `file.fileno()` ---> Returns the integer file descriptor that is used by the underlying implementation to request I/O operations from the operating system.
* `file.isatty()` ---> Returns True if the file is connected to a tty(-like) device, else False.
* `next(file)` ---> Returns the next line from the file each time it is being called.
* `file.read([size])` ---> Reads at most size bytes from the file (less if the read hits EOF before obtaining size bytes).
* `file.readline([size])` ---> Reads one entire line from the file. A trailing newline character is kept in the string.
* `file.readlines([sizehint])` ---> Reads until EOF using readline() and return a list containing the lines. If the optional sizehint argument is present, instead of reading up to EOF, whole lines totalling approximately sizehint bytes (possibly after rounding up to an internal buffer size) are read.
* `file.seek(offset[, whence])` ---> Sets the file's current position
* `file.tell()` ---> Returns the file's current position
* `file.truncate([size])` ---> Truncates the file's size. If the optional size argument is present, the file is truncated to (at most) that size.
* `file.write(str)` ---> Writes a string to the file. There is no return value.
* `file.writelines(sequence)` ---> Writes a sequence of strings to the file. The sequence can be any iterable object producing strings, typically a list of strings.

## Python 3 - OS File/Directory Methods
Update Later https://www.tutorialspoint.com/python3/os_file_methods.htm

## Python 3 - Exceptions Handling
* `try` -> contains the code with possible exceptions raised.
* `except` -> code inside except block executed when an exception is raised (expected exception or unexpected exception)
* `else` -> code inside else will be executed, when no exception is raised.
* `finally` -> code inside finally will be executed in all cases.
* `raise` -> raises exception.

please refer https://thepythonguru.com/python-exception-handling/, https://www.tutorialspoint.com/python3/python_exceptions.htm.

In [39]:
class My_Own_Custom_Exception_Class(RuntimeError):
    def __init__(self, some_arg):
        self.some_arg = some_arg

try:
    raise My_Own_Custom_Exception_Class("My_custom_message")
    raise ZeroDivisionError("You silly,...")
    pass
except ZeroDivisionError as Shammy:
    print("exception caught:" + str(Shammy))
except My_Own_Custom_Exception_Class as custom_var_for_exception:
    print("exception caught:" + str(custom_var_for_exception))
except Exception:
    print("Just Exception")
except:
    print("Any Exception")
else:
    print("No Exception")
finally:
    print("No matter what, this will be printed")

exception caught:My_custom_message
No matter what, this will be printed


## Python 3 - Object Oriented

In [91]:
class CustomClassName:
    'This is doc_string, ie documentation string'
    class_variable = 10
    
    def function_definition_for_name(self, name):
        self.name = name
        
    def function_definition_for_age(age):
        self.age = age
        
    def print_class_name(self):
        print(self.__class__.__name__)
        
    def __del__(self):
        print("This object is being destroyed by Garbage collector")

ref_var_1 = CustomClassName()
ref_var_2 = CustomClassName()

ref_var_1.function_definition_for_name(name=25)
ref_var_1.name

ref_var_1.class_variable = ref_var_1.class_variable + 3
print(ref_var_1.class_variable)
print(ref_var_2.class_variable)
CustomClassName.class_variable = 1777
print(CustomClassName.class_variable)
print("Check these values")
print(ref_var_1.class_variable)
print(ref_var_2.class_variable)
print("test_")
ref_var_1.print_class_name()
print(CustomClassName.__doc__)
ref_var_1_duplicate = ref_var_1
del ref_var_1
del ref_var_1_duplicate
del ref_var_2

13
10
1777
Check these values
13
1777
test_
CustomClassName
This is doc_string, ie documentation string
This object is being destroyed by Garbage collector
This object is being destroyed by Garbage collector


`Once the values of class variables are changed/modified then those values are not changed by changing the class_variables by calling class`
`__del__ function is executed at time of garbage collection. Garbage collection will be triggered if only all the raferences are deleted.`

* `getattr`
* `setattr`
* `hasattr`
* `delattr`

### Built in class attributes.
* `class.__dict__`
* `class.__doc__`
* `class.__name__`
* `class.__module__` --> Module name in which the class is defined. This attribute is "__main__" in interactive mode.
* `class.__bases__` --> A possibly empty tuple containing the base classes, in the order of their occurrence in the base class list.

### Class Inheritance
* `class SubClassName (ParentClass1[, ParentClass2, ...]):`
* `____'Optional class documentation string'`
* `____class_suite`

When you have inherited the methods from multiple classes, if same method is defined in both the parent classes, the first one precedes over the other in inheritence.

In [98]:
class Parent_One:
    def parent_one_method(self):
        print("parent one method")
    def common_parents_method(self):
        print("common_parents_method from parent one")

class Parent_Two:
    def parent_two_method(self):
        print("parent two method")
    def common_parents_method(self):
        print("common_parents_method from parent two")
        
class Child_From_Parents(Parent_Two, Parent_One):
    def child_method(self):
        print("child method")
        
child = Child_From_Parents()
child.parent_one_method()
child.parent_two_method()
child.common_parents_method()

parent one method
parent two method
common_parents_method from parent two


* `issubclass(sub, sup)`
* `isinstance(obj, Class)`

### Base Overloading Methods
* `__init__` ---> Class constructor
* `__del__` ---> Class Destructor
* `__repr__` ---> Evaluatable string representation
* `__str__` ---> Printable string representation
* `__cmp__` ---> Object comparison

In [102]:
print(child.__repr__())
print(child.__str__())

### Overloading Operators
https://docs.python.org/3/library/operator.html Please refer to this for function names of operators.

In [103]:
class Vector:
   def __init__(self, a, b):
      self.a = a
      self.b = b

   def __str__(self):
      return 'Vector (%d, %d)' % (self.a, self.b)
   
   def __add__(self,other):
      return Vector(self.a + other.a, self.b + other.b)

v1 = Vector(2,10)
v2 = Vector(5,-2)
print (v1 + v2)

Vector (7, 8)


### Data Hiding
An object's attributes may or may not be visible outside the class definition. You need to name attributes with a double underscore prefix, and those attributes then will not be directly visible to outsiders.

In [107]:
class Shammy_Thanuja:
    __some_hidden_variable = 78
    def get_hidden_variables(self):
        hidden_var = self.__some_hidden_variable
        return hidden_var
    
var_1 = Shammy_Thanuja()
print(var_1.get_hidden_variables())
try:
    print(var_1.__some_hidden_variable)
except:
    print("You cant access the hidden variables directly")

78
You cant access the hidden variables directly


## Python 3 - Regular Expressions
A regular expression is a special sequence of characters that helps you match or find other strings or sets of strings, using a specialized syntax held in a pattern
There are various characters, which would have special meaning when they are used in regular expression. To avoid any confusion while dealing with regular expressions, we would use Raw Strings as r'expression'.
The module re provides full support for Perl-like regular expressions in Python.

### Basic Patterns that match single characters

* `A, a, X, 9 <, etc` ---> ordinary chanracters that match exactly themselves.
* `.` ---> matches any single character except newline '\n'
* `\w` ---> matches a "word" character: a letter or digit or underbar [a-zA-Z0-9_].
* `\W` ---> matches any non-word character.
* `\b` ---> boundary between word and non-word
* `\s` ---> matches a single whitespace character -- space, newline, return, tab
* `\S` ---> matches any non-whitespace character.
* `\t, \n, \r` ---> Tab, newline, return.
* `\d` ---> Decimal digit.
* `^` ---> matches start of the string
* `$` ---> match the end of the string
* `\` ---> inhibit the "specialness" of a character.

### Compilation flags

Compilation flags let you modify some aspects of how regular expressions work. Flags are available in the re module under two names, a long name such as IGNORECASE and a short, one-letter form such as I.

* `ASCII, A` ---> Makes several escapes like \w, \b, \s and \d match only on ASCII characters with the respective property.
* `DOTALL, S` ---> Make, match any character, including newlines
* `IGNORECASE, I` ---> Do case-insensitive matches
* `LOCALE, L` ---> Do a locale-aware match
* `MULTILINE, M` ---> Multi-line matching, affecting ^ and $
* `VERBOSE, X (for ‘extended’)` ---> Enable verbose REs, which can be organized more cleanly and understandably

### The match Function
This function attempts to match RE _pattern_ to _string_ with optional _flags_.

`re.match(pattern, string, flags = 0)`

* `pattern` ---> This is the regular expression to be matched.
* `string` ---> This is the string, which would be searched to match the pattern at the beginning of string.
* `flags` ---> You can specify different flags using bitwise OR (|). These are modifiers, which are listed in the table below.

The re.match function returns a match object on success, None on failure. We usegroup(num) or groups() function of match object to get matched expression.
* `group(num = 0)` ---> This method returns entire match (or specific subgroup num)
* `groups()` ---> This method returns all matching subgroups in a tuple (empty if there weren't any)

In [128]:
import re 

string_to_be_matched = """Wikipedia is a free online encyclopedia, created and edited by volunteers around the world and hosted by the Wikimedia Foundation."""

matchObj = re.match( r'(.*)online(.*)', string_to_be_matched, re.M|re.I)

if matchObj:
    print(matchObj.groups())
    print(matchObj.group())
    print(str(matchObj.groupdict))
else:
    print("Nothing matched")

('Wikipedia is a free ', ' encyclopedia, created and edited by volunteers around the world and hosted by the Wikimedia Foundation.')
Wikipedia is a free online encyclopedia, created and edited by volunteers around the world and hosted by the Wikimedia Foundation.
<built-in method groupdict of re.Match object at 0x10cfdf270>


### The search Function
This function attempts to search RE _pattern_ to _string_ with optional _flags_.

`re.search(pattern, string, flags = 0)`

* `pattern` ---> This is the regular expression to be searched.
* `string` ---> This is the string, which would be searched to match the pattern at the beginning of string.
* `flags` ---> You can specify different flags using bitwise OR (|). These are modifiers, which are listed in the table below.

The re.search function returns a search object on success, None on failure. We usegroup(num) or groups() function of match object to get matched expression.
* `group(num = 0)` ---> This method returns entire search (or specific subgroup num)
* `groups()` ---> This method returns all searching subgroups in a tuple (empty if there weren't any)

### search vs match
Python offers two different primitive operations based on regular expressions: match checks for a match only at the beginning of the string, while search checks for a match anywhere in the string (this is what Perl does by default).

### Search and Replace
`re.sub(pattern, repl, string, max=0)`

This method replaces all occurrences of the RE pattern in string with repl, substituting all occurrences unless max is provided. This method returns modified string.

In [129]:
import re

phone = "2004-959-559 # This is Phone Number"

# Delete Python-style comments
num = re.sub(r'#.*$', "", phone)
print ("Phone Num : ", num)

# Remove anything other than digits
num = re.sub(r'\D', "", phone)    
print ("Phone Num : ", num)

Phone Num :  2004-959-559 
Phone Num :  2004959559


### Regular Expression Modifiers: Option Flags
* `re.I` ---> Performs case-insensitive matching.
* `re.L` ---> Interprets words according to the current locale. This interpretation affects the alphabetic group (\w and \W), as well as word boundary behavior (\b and \B).
* `re.M` ---> Makes $ match the end of a line (not just the end of the string) and makes ^ match the start of any line (not just the start of the string).
* `re.S` ---> Makes a period (dot) match any character, including a newline.
* `re.U` ---> Interprets letters according to the Unicode character set. This flag affects the behavior of \w, \W, \b, \B.
* `re.X` ---> Permits "cuter" regular expression syntax. It ignores whitespace (except inside a set [] or when escaped by a backslash) and treats unescaped # as a comment marker.

### Regular Expression Examples

Example | Description | Example | Description 
--------|-------------|---------|------------
Python | Matches Python | . | Match any character except newline
[Pp]ython | Matches for python or Python | \d | Match a digit: [0-9]
rub[ye] | Match "ruby" or "rube" | \D | Match a nondigit: [^0-9]
[aeiou] | Match any one lowercase vowel | \s | Match a whitespace character: [ \t\r\n\f]
[0-9] | Match any digit; same as [0123456789] | \S | Match nonwhitespace: [^ \t\r\n\f]
[a-z] | Match any lowercase ASCII letter | \w | Match a single word character: [A-Za-z0-9_]
[A-Z] | Match any uppercase ASCII letter | \W | Match a nonword character: [^A-Za-z0-9_]
[a-zA-Z0-9] | Match any alphanumeric | <.\*> | Greedy repetition: matches "< python > perl >"
[^aeiou] | Match anything other than a lowercase vowel | <.\*?> | Nongreedy: matches '< python >' in "< python > perl >"
[^0-9] | Match anything other than a digit | \D\d+ | No group: + repeats \d
ruby? | Match "rub" or "ruby": the y is optional | (\D\d)+ | Grouped: + repeats \D\d pair
ruby* | Match "rub" plus 0 or more ys | ([Pp]ython(,)?)+ | Match "Python", "Python, python, python", etc.
ruby+ | Match "rub" plus 1 or more ys | ([Pp])ython&\1ails | Match python&pails or Python&Pails
\d{3} | Match exactly 3 digits | (['"])[^\1]\*\1 | Single or double-quoted string. \1 matches whatever the 1st group matched. \2 matches whatever the 2nd group matched, etc.
\d{3,} | Match 3 or more digits | python\|perl | Match "python" or "perl"
\d{3,5} | Match 3, 4, or 5 digits | rub(y\|le) | Match "ruby" or "ruble"
^Python | Match "Python" at the start of a string or internal line | Python(!+\|\?) | "Python" followed by one or more ! or one ?
Python$ | Match "Python" at the end of a string or line | \\APython | Match "Python" at the start of a string
Python\\Z | Match "Python" at the end of a string | \\bPython\\b | Match "Python" at a word boundary
\\brub\\B | \\B is nonword boundary: match "rub" in "rube" and "ruby" but not alone | Python(?=!) | Match "Python", if followed by an exclamation point.
Python(?!!) | Match "Python", if not followed by an exclamation point.

**https://www.tutorialspoint.com/python3/python_reg_expressions.htm for more examples**

In [140]:

some_pattern = r"(.*)[Pp]ython(.*)"

In [141]:
some_string = """Python is an interpreted, high-level, general-purpose programming language. Created by Guido van Rossum and first released in 1991, Python's design philosophy emphasizes code readability with its notable use of significant whitespace. Its language constructs and object-oriented approach aim to help programmers write clear, logical code for small and large-scale projects.[27] Python is dynamically typed and garbage-collected. It supports multiple programming paradigms, including procedural, object-oriented, and functional programming. Python is often described as a "batteries included" language due to its comprehensive standard library.[28] Python was conceived in the late 1980s as a successor to the ABC language. Python 2.0, released 2000, introduced features like list comprehensions and a garbage collection system capable of collecting reference cycles. Python 3.0, released 2008, was a major revision of the language that is not completely backward-compatible, and much Python 2 code does not run unmodified on Python 3. Due to concern about the amount of code written for Python 2, support for Python 2.7 (the last release in the 2.x series) was extended to 2020. Language developer Guido van Rossum shouldered sole responsibility for the project until July 2018 but now shares his leadership as a member of a five-person steering council.[29][30][31] Python interpreters are available for many operating systems. A global community of programmers develops and maintains CPython, an open source[32] reference implementation. A non-profit organization, the Python Software Foundation, manages and directs resources for Python and CPython development."""

In [142]:
match_obj = re.match(some_pattern, some_string)
print(match_obj.groups())

('Python is an interpreted, high-level, general-purpose programming language. Created by Guido van Rossum and first released in 1991, Python\'s design philosophy emphasizes code readability with its notable use of significant whitespace. Its language constructs and object-oriented approach aim to help programmers write clear, logical code for small and large-scale projects.[27] Python is dynamically typed and garbage-collected. It supports multiple programming paradigms, including procedural, object-oriented, and functional programming. Python is often described as a "batteries included" language due to its comprehensive standard library.[28] Python was conceived in the late 1980s as a successor to the ABC language. Python 2.0, released 2000, introduced features like list comprehensions and a garbage collection system capable of collecting reference cycles. Python 3.0, released 2008, was a major revision of the language that is not completely backward-compatible, and much Python 2 code

## An Intro to Threading in Python
Reference: https://realpython.com/intro-to-python-threading/